### ISSUE TITLE
restart too slow and too huge memory footprint

### ISSUE THREAD

**Issue by zhyhang** [pilosa/pilosa#1807]
## Environment
- CentOS7.1 Linux Kernel 3.10.0-862.el7.x86_64
- pilosa v1.1
- single instance
- 384GB memory
- pcie ssd 3.0TB
- **key index**
- **519,870,337 columns**
- **about 50,000 rows**
- **.key file(translation map file) size about  100GB**
- memory usage: **RES 190.4GB and SHR 99.4GB**

## Suffering
- restart that pilosa taking 2 hours, I found only one thread load data. 
- too many memory footprint, about

**Comment by jaffee** [comment]
related #605

**Comment by jaffee** [comment]
I don't think we've done much testing with a key translation file that large, so that could be a factor

**Comment by jaffee** [comment]
@zhyhang how large are your pilosa data files? (not the key translation file)

I would also like to see the golang heap profile at the end of the two hours. you can obtain this by `curl pilosa:10101/debug/pprof/heap > startup.prof`. If you could send that file it would be very helpful. thank you!

**Comment by zhyhang** [comment]
### heap dump, please see the attachment
[startup.tar.gz](https://github.com/pilosa/pilosa/files/2719289/startup.tar.gz)

### following is start log snippet:
2019/01/01 16:33:35 open server
2019/01/01 18:35:42 WARNING: Tried to set open file limit to 262144, but it is 65535. You may consider running "sudo ulimit -n 262144" before starting Pilosa to avoid "too many open files" error. See https://www.pilosa.com/docs/latest/administration/#open-file-limits for more information.
2019/01/01 18:35:42 open holder path: /ssd/cache/pilosa/data
2019/01/01 18:35:42 opening index: adc
2019/01/01 18:35:42 opening index: wx
2019/01/01 18:35:42 opening index: mac
2019/01/01 18:35:42 opening index: pyc
2019/01/01 18:35:54 opening index: did
2019/01/01 18:35:59 opening index: ph
2019/01/01 18:35:59 opening index: mb
2019/01/01 18:35:59 opening index: wb
2019/01/01 18:35:59 open holder: complete
2019/01/01 18:35:59 received state READY (ae2da9e0-6c28-4b65-bcb5-a59415009b1a)
2019/01/01 18:35:59 change cluster state from STARTING to NORMAL on ae2da9e0-6c28-4b65-bcb5-a59415009b1a
2019/01/01 18:35:59 listening as http://192.168.152.192:11101
2019/01/01 18:35:59 holder sync monitor initializing (10m0s interval)
2019/01/01 18:35:59 Pilosa is currently configured to send small diagnostics reports to our team every 1h0m0s. More information here: https://www.pilosa.com/docs/latest/administration/#diagnostics

**Comment by jaffee** [comment]
based on the timing of those log messages, it would appear that the long load time is occurring during the opening of the translate file. The memory profile with most memory being allocated by TranslateFile.applyEntry.

Will have to dig a bit more to determine what this means exactly and what we can do about it.

**Comment by zhyhang** [comment]
@jaffee Thanks!

**Comment by zhyhang** [comment]
@jaffee I would like divide the big instance to some smaller instances. Is It can reduce the total memory usage which occupied by .key file and bitmap data?

**Comment by jaffee** [comment]
@zhyhang it will help with the bitmap data, but the translation file will be replicated to every node. Translation is something we hope to make more scalable in the future. 

I'm fairly confident that we can improve the startup time that you are seeing significantly, but will have to spend some more time with it.

**Comment by zhyhang** [comment]
@jaffee I believe that!

**Comment by benbjohnson** [comment]
@jaffee The translation file is a linear log so it's not easy to parallelize. Splitting it so it's per-index would help (see https://github.com/pilosa/pilosa/issues/1976) since they could load independently.

However, a separate issue is that we are storing a [reverse lookup from id-to-offset](https://github.com/benbjohnson/pilosa/blob/7ff684c193dd6c0fa71b8ea9b0d0141e17adc3dd/translate.go#L890). It's a map of uint64-to-uint64 so with 520M columns times 16 bytes that's 8GB and that doesn't factor in the overhead of the map. We may want to persist that to disk. We also maintain a [list of elements](https://github.com/benbjohnson/pilosa/blob/7ff684c193dd6c0fa71b8ea9b0d0141e17adc3dd/translate.go#L896) which adds substantial memory overhead.

We may want to persist the whole index to disk using Badger or Bolt. We will definitely take a read & write performance hit compared to the RHH implementation we have now but it would scale up much further.

**Comment by jaffee** [comment]
Moving this to in-progress since a rewrite of the translate store is underway in #2042

**Comment by travisturner** [comment]
this was addressed by #2042
